src/montgomery.rs

```diff
@@ -193,6 +193,81 @@ pub(crate) fn elligator_encode(r_0: &FieldElement) -> MontgomeryPoint {
 }
 
+/// Perform the inverse Elligator2 mapping from a Montgomery point to
+/// field element. This algorithm is based on [Elligator:
+/// Elliptic-curve points indistinguishable from uniform random
+/// strings][elligator], Section 5.3.
+///
+/// This function is a partial right inverse of `elligator_encode`: if
+/// `elligator_decode(&p, sign) == Some(fe)` then `elligator_encode(&fe)
+/// == p`.
+///
+/// This function does _not_ operate in constant time: if `point`
+/// cannot be mapped to a field element, the function exits early. In
+/// typical usage this function is combined with rejection sampling
+/// and called repeatedly until a mappable point is found.
+///
+/// Note: the output field elements of this function are uniformly
+/// distributed among the nonnegative field elements, but only if the
+/// input points are also uniformly distributed among all points of
+/// the curve. In particular, if the inputs are only selected from
+/// members of the prime order group, then the outputs are
+/// distinguishable from random.
+///
+/// # Inputs
+///
+/// * `point`: the \\(u\\)-coordinate of a point on the curve. Not all
+/// points map to field elements.
+///
+/// * `v_is_negative`: true if the \\(v\\)-coordinate of the point is negative.
+///
+/// # Returns
+///
+/// Either `None`, if the point couldn't be mapped, or `Some(fe)` such
+/// that `fe` is nonnegative and `elligator_encode(&fe) == point`.
+///
+/// [elligator] https://elligator.cr.yp.to/elligator-20130828.pdf
+///
+#[allow(unused)]
+pub(crate) fn elligator_decode(point: &MontgomeryPoint, v_is_negative: Choice) -> Option<FieldElement> {
+    let one = FieldElement::one();
+    let u = FieldElement::from_bytes(&point.to_bytes());
+    let u_plus_A = &u + &MONTGOMERY_A;
+    let uu_plus_uA = &u * &u_plus_A;
+
+    // Condition: u is on the curve
+    let vv = &(&u * &uu_plus_uA) + &u;
+    let (u_is_on_curve, _v) = FieldElement::sqrt_ratio_i(&vv, &one);
+    if (!bool::from(u_is_on_curve)) {
+        return None
+    }
+
+    // Condition: u != -A
+    if (u == MONTGOMERY_A_NEG) {
+        return None
+    }
+
+    // Condition: -2u(u+A) is a square
+    let uu2_plus_uA2 = &uu_plus_uA + &uu_plus_uA;
+    // We compute root = sqrt(-1/2u(u+A)) to speed up the calculation.
+    // This is a square if and only if -2u(u+A) is.
+    let (is_square, root) = FieldElement::sqrt_ratio_i(&FieldElement::minus_one(),
+                                                       &uu2_plus_uA2);
+    if (!bool::from(is_square | root.is_zero())) {
+        return None;
+    }
+
+    // if !v_is_negative: r = sqrt(-u / 2(u + a)) = root * u
+    // if  v_is_negative: r = sqrt(-(u+A) / 2u)   = root * (u + A)
+    let add = FieldElement::conditional_select(&u, &u_plus_A, v_is_negative);
+    let r = &root * &add;
+
+    // Both r and -r are valid results. Pick the nonnegative one.
+    let result = FieldElement::conditional_select(&r, &-&r, r.is_negative());
+
+    return Some(result);
```

Comment:
You may omit `return` here in the last line of a function, as in other functions in this file.
```suggestion
    Some(result)
```